docs/design/04-asset-records.md

```diff
@@ -0,0 +1,131 @@
+- Asset records will need to be assigned to nodes (and reference each other
+  at some future date)
+- Some basic "validation" of the files
+
+# New Commands:
+
+There will be three basic commands:
+- `metal asset add TYPE NAME`
+- `metal asset edit NAME`
+- `metal asset delete NAME`
+
+# Adding a new asset record
+
+To make a new asset record, first a blank "template" needs to be loaded from
+repo. Note that these "templates" are not ERB. Instead they are the "YAML"
+that will provide the basis for the asset record. An no point will the
+entire "template" be rendered using ERB (however the individual tags might?).
+
+The template will then be opened using the system editor (see issues). Before
+being saved to the final location in `/var/lib/metalware/assets`. See
+`Editing` for more details on this process.
+
+In short, the validation will occur when the file is saved. This means the
+initial "templates" do not need to be valid `YAML`. Be careful of the file
+extension as it might not be `.yaml` (see issues).
+
+As this is a new record, the `type` is required for the command. This tells
+`metalware` which "template" should be used. The templates should be stored
+under `repo/assets`. The `name` is to be used for saving the asset. So a
+typical command would look like:
+`metal asset add rack rack1`
+
+# Editing an existing record
+As the asset already has a name, the command will look like:
+`metal asset edit rack1`.
+The type is not required as the asset can be retrieved with the name itself.
+
+This will be similar to adding a new record except the old record is used
+instead of the template. Otherwise the following process is the same for both.
+
+The edit process should not be destructive until after the updated file has
+been validated. This means the files can not be edited in place. Instead the
+existing record should be moved to a temporary file which is then edited.
+
+Once the user has finished editing the file and saves it, `metalware` should
+detect the editor has closed and continue on with the process. This should be
+the case when using `POpen.capture3`, however it might need to poll the
+editor `PID` until it is closed.
+
+At this stage, only basic validation of the file is required. The file is
+valid if it can be reloaded as a hash using `Data.load`. If the file is not
+valid, then the process should end with an error and point the user to the
+bad file's temp location. NOTE: Ruby does automatically delete temporary
+files, so special handling maybe required to prevent this.
+
+At this stage re-editing a previously bad file does not need to be supported.
+This can be added in the future.
+
+At the end, the temporary file is moved to its final location.
+
+# Accessing Asset Records and Naming
+
+The asset records will be loaded into the namespace under `alces.assets`.
+They will be accessible by name as both: `alces.assets.<name>` and
+`alces.assets.find_by_name(name)`. The `MetalArray` could be used for this
```

Comment:
Is there a benefit of having the `alces.assets.find_by_name('name')` form as opposed to running `alces.assets.name` directly?

Reply:
The two methods for accessing have slightly different purposes. If you know the name in advanced and it can be hard coded, then `alces.assets.<name>` is preferred.

However if the name is a variable or won't be known at runtime then `send` would have to be used to get the asset. As this could have side affects, a more robust `find_by_name` method is preferred.

So having both methods is ideal depending on the circumstances. This matches the accessing methods used for `alces.groups` and `alces.nodes`.